src/feditest/nodedrivers/mastodon/api.py

```diff
@@ -0,0 +1,295 @@
+
+from feditest import nodedriver
+from feditest.protocols import Node, NodeDriver
+from feditest.protocols.activitypub import ActivityPubNode
+
+
+def _dereference(uri: str) -> dict:
+    return httpx.get(
+        uri,
+        follow_redirects=True,
+        verify=False,
+        headers={"Accept": "application/activity+json"},
+    ).json()
+
+
+class Actor:
+    """Encapsulates account/actor related information. For a Mastodon "actor"
+    it contains a reference to an API wrapper and the account data. For an actor
+    without a Mastodon access_token, it has the uri and the AP actor document.
+    """
+
+    def __init__(self, role: str, hostname: str, params: str):
+        if access_token := params.get("access_token"):
+            self.api = Mastodon(
+                api_base_url=f"https://{hostname}", access_token=access_token
+            )
+            self.data = self.api.account_verify_credentials()
+        else:
+            self.api = None
+            self.data = None
+        self.role = role
+        self.uri = params["uri"]
+        self.actor = _dereference(self.uri)
+
+    @property
+    def is_mastodon(self):
+        return self.api is not None
+
+
+class MastodonNode(ActivityPubNode):
+    """
+    Supported Parameters:
+        {
+            "hostame": "somehost.org",
+            "inbox_wait": {"retry_count" 5, "retry_interval": 1}, # optional
+            "follow_wait": {"retry_count" 30, "retry_interval": 1}, # optional
+            "actors": {
+                "default_role": "primary_actor",
+                "roles": {
+                    "primary_actor": {
+                        uri: "PRIMARY_ACTOR_URI",
+                        access_token: "ACCESS_TOKEN"
+                    },
+                    "secondary_actor": {
+                        uri: "SECONDARY_ACTOR_URI",
+                    },
+                    "external_actor": {
+                        uri: "EXTERNAL_ACTOR_URI",
+                    }
+                }
+            },
+        }
+
+    Required roles depend on tests.
+    """
+
+    def __init__(
+        self, rolename: str, parameters: dict[str, Any], node_driver: NodeDriver
+    ):
+        super().__init__(rolename, parameters, node_driver)
+        self.actors = [
+            Actor(role_name, self._param("hostname"), role_params)
+            for role_name, role_params in self._param("actors", "roles").items()
+        ]
+        self.actors_by_uri = {a.uri: a for a in self.actors}
+        self.actors_by_role = {a.role: a for a in self.actors}
+        self.default_actor = next(
+            a for a in self.actors if a.role == self._param("actors", "default_role")
+        )
+        self.uri_map: dict[str, dict] = dict()
+
+    def _get_actor_or_default(self, actor_uri: str | None) -> Actor:
+        return self.actors_by_uri[actor_uri] if actor_uri else self.default_actor
+
+    def _param(self, *keys, default: Any = None, required: bool = True):
```

Comment:
This was a puzzler until I realized it is a get over multiple levels of a hierarchy!